pandas/core/indexing.py

```diff
@@ -2195,10 +2196,12 @@ def check_bool_indexer(index: Index, key) -> np.ndarray:
                 "the indexed object do not match)."
             )
         result = result.astype(bool)._values
-    else:
+    elif is_object_dtype(key):
         # key might be sparse / object-dtype bool, check_array_indexer needs bool array
         result = np.asarray(result, dtype=bool)
         result = check_array_indexer(index, result)
+    else:
+        result = check_array_indexer(index, result)
```

Comment:
might be able to simply to check_array_indexer right before returning (iow for all cases) here. (try in a followon)